demos/speculation_over_syscall.cc

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <array>
+#include <cstring>
+#include <fstream>
+#include <iostream>
+
+#ifndef __linux__
+#  error Unsupported OS. Linux required.
+#endif
+
+#ifndef __aarch64__
+#  error Unsupported architecture. ARM64 required.
+#endif
+
+#include <linux/membarrier.h>
+#include <signal.h>
+#include <sys/mman.h>
+#include <sys/syscall.h>
+#include <unistd.h>
+
+#include "cache_sidechannel.h"
+#include "instr.h"
+
+const char *public_data = "Hello, world!";
+const char *private_data = "It's a s3kr3t!!!";
+constexpr size_t PAGE_SIZE = 4096;
```

Comment:
This likely belongs in `instr.h`.

Reply:
PowerPC has page size 64kB. Do you want me to create an ifdef for that? 64kB for PowerPC and 4kB for everything else?

Reply:
> > What syscall do you consider more straightforward and still implementable with the same reliability?
> 
> A few ideas off the top of my head:
> 
> * use `ptrace` to skip past some instructions, but show they were executed speculatively
> * `raise()` (`kill` self)
> * create a new thread that calls `pthread_exit` before performing architecturally-dead reads, then join on that thread and show the reads actually happened speculatively. (Will involve some trickery for the compiler not to optimize out code after `pthread_exit`. May also require going straight to `syscall(SYS_exit, 0);` or even doing [raw threads with `clone`](https://eli.thegreenplace.net/2018/launching-linux-threads-and-processes-with-clone/) if `pthread` adds too much overhead and exhausts the speculation window.)
> 
> > The indirect calls induced by Spectre v2 can never reach the 40 hit boundary from cache_sidechannel.cc unless I intentionally make them to do so
> 
> If we are totally confident this is true, let's add a comment to that effect (so folks know there are sharp edges here and we've taken care not to get false results). But I admit I'd be more comfortable if, knowing there is fragility here (don't ever remove that `+40`!), we wrote different code that avoided it.

raise is a library call, but I can use inline assembly to kill itself (__NR_kill) with a signal that can be captured (e. g. SIGUSR1) and move the PC to the afterspeculation label. That will avoid creating threads and will be probably very robust.